app/src/Command.php

```diff
@@ -0,0 +1,153 @@
+<?php
+
+require_once(__DIR__ . '/Db.php');
+require_once(__DIR__ . '/XmlParser.php');
+
+class Command 
+{
+  private $db;
+  private $samlDisplaynameMapping;
+  private $samlEmailMapping;
+  private $samlGroupMapping;
+  private $samlGeneralUidMapping;
+  private $spX509Cert;
+  private $spPrivateKey;
+  private $xml;
+
+  public function __construct(
+    Db $db,
+    XmlParser $xml,
+    string $samlDisplaynameMapping,
+    string $samlEmailMapping,
+    string $samlGroupMapping,
+    string $samlGeneralUidMapping,
+    string $spX509Cert,
+    string $spPrivateKey
+  ){
+    $this->db = $db;
+    $this->xml = $xml;
+    $this->samlDisplaynameMapping = $samlDisplaynameMapping;
+    $this->samlEmailMapping = $samlEmailMapping;
+    $this->samlGroupMapping = $samlGroupMapping;
+    $this->samlGeneralUidMapping = $samlGeneralUidMapping;
+    $this->spX509Cert = $spX509Cert;
+    $this->spPrivateKey = $spPrivateKey;
+  }
+
+  public function execute()
+  {
+    echo "Start executing syncro script".PHP_EOL;
+    // First fetch the current propertyIds parameter and save it in class property
+    $this->providerIds = $this->db->getProviderIds();
+    // Prepared statement checking if a given entity provider already exists
+    $identityProviderExistsStatement = $this->db->getConnection()->prepare(<<<SQL
+SELECT configkey
+FROM appconfig
+WHERE appid='user_saml'
+AND configkey ~ '^[0-9]*-?idp-entityId$'
+AND configvalue=:entityId
+SQL
+);
+
+    // Iterate other all providers present in the XML file
+    foreach ($this->xml->getEntityDescriptors() as $entityDescriptor) {
+      $entityId = $entityDescriptor->attributes()->entityID;
+      // Execute the prepared statement with the current entity provider
+      $identityProviderExistsStatement->execute(['entityId' => $entityId]);
+
+      // If the entity provider does not exist, create it.
+      if (false === $idpEntityIdConfigKey = $identityProviderExistsStatement->fetchColumn()) {
+        $this->insertIdentityProvider($entityId, $entityDescriptor);
+      } else {
+        // If the entity provider exists, update it.
+        $this->updateIdentityProvider($idpEntityIdConfigKey, $entityId, $entityDescriptor);
+      }
+    }
+  }
+
+  private function updateIdentityProvider(string $idpEntityIdConfigKey, string $entityId, SimpleXMLElement $entityDescriptor) 
+  {
+    echo sprintf('Updating identity provider: %s', $entityId).PHP_EOL;
+    // Retrieve provider id from $idpEntityIdConfigKey
+    preg_match('/^([0-9]*)-?idp-entityId$/', $idpEntityIdConfigKey, $matches);
+    $id = $matches[1] ?: 1;
+    $prefix = $id === '1' ? '' : $id . '-';
+    
+    $this->processIdentityProvider($prefix, $entityId, $entityDescriptor);
+  }
+
+  private function insertIdentityProvider(string $entityId, SimpleXMLElement $entityDescriptor)
+  {
+    echo sprintf('Adding new identity provider: %s', $entityId).PHP_EOL;
+    $id = end($this->providerIds)?: 0;
+    array_push($this->providerIds, ++$id);
```

Comment:
I find the following version easier to read, WDYT ?

```suggestion
    $id = count($this->providerIds) + 1;
    $this->providerIds[] = $id;
```

Reply:
Yes but it is not the same meaning, the array is sorted but an index can be missing and having this kind of list `[1, 2, 4, 5]`

Reply:
Oh, right :+1: